include/Rotation.h

```diff
@@ -0,0 +1,81 @@
+#ifndef ALE_ROTATION_H
+#define ALE_ROTATION_H
+
+#include <memory>
+#include <vector>
+
+namespace ale {
+
+  enum RotationInterpolation {
+    slerp,
+    nlerp
+  };
+
+  class Rotation {
+    public:
+      Rotation();
+      Rotation(double w, double x, double y, double z);
+      Rotation(const std::vector<double>& matrix);
+      Rotation(const std::vector<double>& angles, const std::vector<int>& axes);
+      Rotation(const std::vector<double>& axis, double theta);
+      ~Rotation();
+
+      // Special member functions
+      Rotation(Rotation && other) noexcept;
+      Rotation& operator=(Rotation && other) noexcept;
+
+      Rotation(const Rotation& other);
+      Rotation& operator=(const Rotation& other);
+
+      // Type specific accessors
+      std::vector<double> toQuaternion() const;
+      std::vector<double> toRotationMatrix() const;
+      std::vector<double> toEuler(const std::vector<int>& axes) const;
+      std::pair<std::vector<double>, double> toAxisAngle() const;
+
+      // Generic rotation operations
+      std::vector<double> operator()(const std::vector<double>& vector, const std::vector<double>& av = {0.0, 0.0, 0.0}) const;
+      Rotation inverse() const;
+      Rotation operator*(const Rotation& rightRotation) const;
+      Rotation interpolate(const Rotation& nextRotation, double t, RotationInterpolation interpType) const;
+
+    private:
+      // pimpl
+      class Impl;
```

Comment:
Why make this a separate class?

Reply:
This hides the third party dependency from the API which helps isolate this better.

https://www.geeksforgeeks.org/pimpl-idiom-in-c-with-examples/

Reply:
Mostly, this makes it so that down-stream projects don't need to include the Eigen headers.

Reply:
If we want to not expose the Eigen headers, as far as I know we only require for the Eigen types/functions to exist in the CPP and the interface without the nested class doesn't really need to expose Eigen classes/functions.